### PR TITLE
[typescript] 4.1 template literal type

### PR DESCRIPTION
**Pull request by g-plane**
Implemented another TypeScript 4.1 feature: template literal type.

I'm not sure whether [ts-parser-test-ref](https://github.com/swc-project/ts-parser-test-ref) should be updated or not.

<details>
<summary>Some known failed reference tests</summary>

```
test typescript::reference::tsc/es6/templates/templateStringControlCharacterEscapes01_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringControlCharacterEscapes01/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringControlCharacterEscapes02_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringControlCharacterEscapes02/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringControlCharacterEscapes03/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringControlCharacterEscapes03_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringTermination2/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringTermination2_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringTermination4/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringTermination4_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringTermination5/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringTermination5_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringWhitespaceEscapes2/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringWhitespaceEscapes1/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringWhitespaceEscapes1_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringWhitespaceEscapes2_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringWithBackslashEscapes01/input.ts ... FAILED
test typescript::reference::tsc/es6/templates/templateStringWithBackslashEscapes01_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings01_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings02_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings02_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings01_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings04_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings03_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings04_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings03_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings06_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings08_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings08_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings06_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings09_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings13_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings13_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings09_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings15_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings15_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings16_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInStrings16_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates01_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates02_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates01_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates02_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates03_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates04_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates03_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates04_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates06_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates06_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates08_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates08_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates09_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates13_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates09_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates13_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates15_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates16_ES5/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates15_ES6/input.ts ... FAILED
test typescript::reference::tsc/es6/unicodeExtendedEscapes/unicodeExtendedEscapesInTemplates16_ES6/input.ts ... FAILED
test typescript::reference::tsc/interfaces/interfaceDeclarations/asiPreventsParsingAsInterface01/input.ts ... FAILED
test typescript::reference::tsc/interfaces/interfaceDeclarations/asiPreventsParsingAsInterface02/input.ts ... FAILED
test typescript::reference::tsc/interfaces/interfaceDeclarations/asiPreventsParsingAsInterface03/input.ts ... FAILED
test typescript::reference::tsc/parser/ecmascript5/parserS7.2_A1.5_T2/input.ts ... FAILED
test typescript::reference::tsc/parser/ecmascript5/parserS7.3_A1.1_T2/input.ts ... FAILED
test typescript::reference::tsc/parser/ecmascript5/parserS7.6_A4.2_T1/input.ts ... FAILED
test typescript::reference::tsc/parser/ecmascript5/parserUnicode1/input.ts ... FAILED
test typescript::reference::tsc/scanner/ecmascript5/scannerS7.2_A1.5_T2/input.ts ... FAILED
test typescript::reference::tsc/scanner/ecmascript5/scannerS7.3_A1.1_T2/input.ts ... FAILED
test typescript::reference::tsc/parser/ecmascript5/parserRealSource1/input.ts ... FAILED
test typescript::reference::tsc/scanner/ecmascript5/scannerS7.6_A4.2_T1/input.ts ... FAILED
```

</details>

### CODE REVIEW

File: ecmascript/ast/src/typescript.rs [hunk 1]
```diff
@@ -741,7 +741,18 @@ pub enum TsLit {
     BigInt(BigInt),
 
     #[tag("TemplateLiteral")]
-    Tpl(Tpl),
+    Tpl(TsTplLitType),
```
Comment (kdy1) [inline ecmascript/ast/src/typescript.rs:744]: Why did you change this to `TsTuplLitType`?

**Review by kdy1 — COMMENTED** [review]
I'm not sure if this fix is correct.

File: ecmascript/ast/src/typescript.rs (hunk 1 above)
Comment (g-plane) [reply ecmascript/ast/src/typescript.rs:744]: `Tpl` is for expression-based template literal, not for type-based, eg: `Tpl` should appear as RValue.

Comment (kdy1) [reply ecmascript/ast/src/typescript.rs:744]: I got it, thanks for the explanation.

**Review by kdy1 — APPROVED** [review]
LGTM, thanks!

**Comment by kdy1** [comment]
CI is killed by oom killer.
I'll merge this after reducing peek memory of the parser testing suite.
